### ISSUE TITLE
PTHREAD error when compiling htslib

### ISSUE THREAD

**Issue by jdmontenegro** [samtools/htslib#420]
$> autoheader
$> autoconf
$> ./configure
$> make
(...)
echo '#define HTS_VERSION "1.3.2-129-gdaae2ea"' > version.h
gcc -g -O2 -I.  -c -o hts.o hts.c
gcc -g -O2 -I.  -c -o md5.o md5.c
gcc -g -O2 -I.  -c -o probaln.o probaln.c
gcc -g -O2 -I.  -c -o realn.o realn.c
gcc -g -O2 -I.  -c -o regidx.o regidx.c
gcc -g -O2 -I.  -c -o sam.o sam.c
gcc -g -O2 -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c
gcc -g -O2 -I.  -c -o vcf_sweep.o vcf_sweep.c
gcc -g -O2 -I.  -c -o tbx.o tbx.c
gcc -g -O2 -I.  -c -o thread_pool.o thread_pool.c
thread_pool.c: In function ‘hts_tpool_init’:
thread_pool.c:658:38: error: ‘PTHREAD_MUTEX_RECURSIVE’ undeclared (first use in this function)
     pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
                                      ^
thread_pool.c:658:38: note: each undeclared identifier is reported only once for each function it appears in
make: **\* [thread_pool.o] Error 1

cannot build the development version of htslib in the server:

$> uname -a

Linux zeus 3.0.101-80-default #1 SMP Fri Jul 15 14:30:41 UTC 2016 (eb2ba81) x86_64 x86_64 x86_64 GNU/Linux

Any ideas how to fix it?

Cheers,

**Comment by jkbonfield** [comment]
Hmm this define has been around since 2008 at least:

http://pubs.opengroup.org/onlinepubs/9699919799/

How old is that linux distro? 3.0 is 2011ish, so I'd think it's new enough so maybe it's some compiler defines we miss.

Does this compile if you define _POSIX_C_SOURCE?  If not, try _POSIX_SOURCE or if desparate _GNU_SOURCE.  We need to know why this function/argument is apparently missing on your OS.

It's possible it can be rewritten to avoid this, but a recursive lock was the least complex solution.

**Comment by jdmontenegro** [comment]
Hi James,

The system administrator managed to get it compiled. I asked him and he
mentioned the need to add an extra flag to gcc during compilation.
Apparently it was a known problem with pthreads in gcc with the system
itself. He did not mentioned what flag he used though, but I am guessing
"-pthreads"

Cheers,

Juan Montenegro

2016-10-06 17:01 GMT+10:00 James Bonfield <notifications@github.com>:

**Comment by jkbonfield** [comment]
That's an interesting point and probably a mistake of our configure script.  I see from your compiler output that it didn't build with -pthreads.  This is an error I believe as the gcc manual claims it is used for both the C preprocessor as well as linking.

Thanks for the bug report.

**Comment by jmarshall** [comment]
I am aware of the theoretical need to compile as well as link with `-pthread` but had not previously met a system where it makes a difference!  Indeed we should probably be probing for pthread details, and it would be good to fix the problem underlying #255 at the same time.  So I look forward to @jkbonfield's pull request :smile:

It would still be interesting to hear more details of your OS and pthread implementation.

Articles like this [expert's rant](http://www.zaval.org/resources/library/butenhof1.html) suggest that recursive mutexes may not be the best choice in general…

**Comment by jkbonfield** [comment]
Agreed recursive mutex isn't ideal, but it became necessary (or something emulating it manually) once we started to allow sharing of queues and pools together as one can call the other and vice versa.  The alternative is a non recursive mutex and a manual counter, but why reinvent the wheel plus it doesn't address any of the issues mentioned in that article, which basically implies our fundamental design is flawed.  Maybe, but it's *complex* and took ages to get this (hopefully) safe.

With regards to -pthread being necessary, we're all guessing really.  @jdmontenegro is there any chance you can confirm whether that was the necessary tweak to get it to compile for you?  Also which OS and version is this system running? Uname was quite vague on that front!

**Comment by jdmontenegro** [comment]
Hi,
I talked to the system manager and he mentioned he used this flag:

–D_GNU_SOURCE

this seems to be a problem with some systems only. Unfortunately I have not
been able to collect any more information about the OS itself. All I know
is what I got from the uname command. I believe they were using CENTOS
until not long ago, but that is no longer the case.

Cheers,

Juan Montenegro

2016-10-06 19:03 GMT+10:00 James Bonfield <notifications@github.com>:
